operator/bucketcontroller/webhook_test.go

```diff
@@ -0,0 +1,193 @@
+		t.Run(name, func(t *testing.T) {
+			bucket := &exoscalev1.Bucket{
+				ObjectMeta: metav1.ObjectMeta{Name: "bucket"},
+				Spec: exoscalev1.BucketSpec{
+					ResourceSpec: xpv1.ResourceSpec{
+						ProviderConfigReference: &xpv1.Reference{Name: tc.providerName},
+					},
+					ForProvider: exoscalev1.BucketParameters{BucketName: "bucket"},
+				},
+			}
+			v := &BucketValidator{log: logr.Discard()}
+			err := v.ValidateCreate(nil, bucket)
+			if tc.expectedError != "" {
+				assert.EqualError(t, err, tc.expectedError)
+			} else {
+				require.NoError(t, err)
+			}
+		})
+	}
+}
+
+func TestBucketValidator_ValidateUpdate_PreventBucketNameChange(t *testing.T) {
+	tests := map[string]struct {
+		newBucketName string
+		oldBucketName string
+		expectedError string
+	}{
+		"GivenNoNameInStatus_WhenNoNameInSpec_ThenExpectNil": {
+			oldBucketName: "",
+			newBucketName: "",
+		},
+		"GivenNoNameInStatus_WhenNameInSpec_ThenExpectNil": {
+			oldBucketName: "",
+			newBucketName: "my-bucket",
+		},
+		"GivenNameInStatus_WhenNameInSpecSame_ThenExpectNil": {
+			oldBucketName: "my-bucket",
+			newBucketName: "my-bucket",
+		},
+		"GivenNameInStatus_WhenNameInSpecEmpty_ThenExpectNil": {
+			oldBucketName: "bucket",
+			newBucketName: "", // defaults to metadata.name
+		},
+		"GivenNameInStatus_WhenNameInSpecDifferent_ThenExpectError": {
+			oldBucketName: "my-bucket",
+			newBucketName: "different",
+			expectedError: `a bucket named "my-bucket" has been created already, you cannot rename it`,
+		},
+	}
+	for name, tc := range tests {
+		t.Run(name, func(t *testing.T) {
+			oldBucket := &exoscalev1.Bucket{
+				ObjectMeta: metav1.ObjectMeta{Name: "bucket"},
+				Spec: exoscalev1.BucketSpec{
+					ForProvider:  exoscalev1.BucketParameters{BucketName: tc.oldBucketName},
+					ResourceSpec: xpv1.ResourceSpec{ProviderConfigReference: &xpv1.Reference{Name: "provider-config"}},
+				},
+				Status: exoscalev1.BucketStatus{AtProvider: exoscalev1.BucketObservation{BucketName: tc.oldBucketName}},
+			}
+			newBucket := &exoscalev1.Bucket{
+				ObjectMeta: metav1.ObjectMeta{Name: "bucket"},
+				Spec: exoscalev1.BucketSpec{
+					ForProvider:  exoscalev1.BucketParameters{BucketName: tc.newBucketName},
+					ResourceSpec: xpv1.ResourceSpec{ProviderConfigReference: &xpv1.Reference{Name: "provider-config"}},
+				},
+			}
+			v := &BucketValidator{log: logr.Discard()}
+			err := v.ValidateUpdate(nil, oldBucket, newBucket)
+			if tc.expectedError != "" {
+				assert.EqualError(t, err, tc.expectedError)
+			} else {
+				require.NoError(t, err)
+			}
+		})
+	}
+}
+
+func TestBucketValidator_ValidateUpdate_RequireProviderConfig(t *testing.T) {
+	tests := map[string]struct {
+		providerName  string
+		expectedError string
+	}{
+		"GivenProviderName_ThenExpectNoError": {
+			providerName: "provider-config",
+		},
+		"GivenNoProviderName_ThenExpectError": {
+			providerName:  "",
+			expectedError: `.spec.providerConfigRef.name is required`,
+		},
+	}
```

Comment:
Sorry for not spotting it earlier, but the provider config reference can also be possibly nil and we should check for that.

Alternatively, I think it may start making sense to extract some individual checks into their own function, so that we don't have to test them twice (once in Create and once in Update)

Reply:
In this case it makes sense to group them by object(only 1 or 2 strings) as long as the object is not complex.